### ISSUE TITLE
minimize dependencies

### ISSUE THREAD

**Issue by jastice** [scalameta/lsp4s#8]
Followup issue for https://github.com/scalameta/metals/issues/285

Logback is replaced with scribe, but an effort to replace circe and cats in #4 was abandoned. For distribution in clients, a smaller library footprint is very desirable.

Olaf suggested alot of the bulk can be removed by pruning unused cats implicits.

**Comment by gabro** [comment]
At some point we had the idea of using proguard for it, maybe @jvican has some experience with it?

**Comment by olafurpg** [comment]
I think we might want to wait and see if there will be a bsp4j (similar to lsp4j). That would likely be a much slimmer artifact and avoids us to jump through hoops to reduce the weight in lsp4s. I agree it would be nice if lsp4s brought in fewer dependencies but it's not something I will personally prioritize in the coming months.

**Comment by olafurpg** [comment]
The situation has improved after #11 is merged so that shapeless, pprint and enumeratum have been removed the the dependencies of the `jsonrpc` module.

**Comment by jastice** [comment]
That's great! Personally, I'm basically happy to have cats in the project so probably won't be putting effort into classpath mungery.

**Comment by olafurpg** [comment]
Sweet, let's close this once #11 is merged then 👍

**Comment by olafurpg** [comment]
v0.2.0 has been released and the total footprint of jsonrpc is now at 12mb

```
coursier fetch org.scalameta:jsonrpc_2.12:0.2.0  -r sonatype:releases | grep -v "org/scala-lang" | xargs du -hc
 32K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/com/outr/perfolation_2.12/1.0.2/perfolation_2.12-1.0.2.jar
8.0K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/org/typelevel/cats-macros_2.12/1.1.0/cats-macros_2.12-1.1.0.jar
184K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/org/jctools/jctools-core/2.0.1/jctools-core-2.0.1.jar
100K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/io/circe/circe-derivation_2.12/0.9.0-M4/circe-derivation_2.12-0.9.0-M4.jar
8.0K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/com/outr/scribe-macros_2.12/2.5.0/scribe-macros_2.12-2.5.0.jar
112K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/io/monix/monix-types_2.12/2.3.3/monix-types_2.12-2.3.3.jar
296K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/io/monix/monix-eval_2.12/2.3.3/monix-eval_2.12-2.3.3.jar
2.1M	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/io/monix/monix-reactive_2.12/2.3.3/monix-reactive_2.12-2.3.3.jar
4.0K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/org/reactivestreams/reactive-streams/1.0.0/reactive-streams-1.0.0.jar
872K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/io/circe/circe-core_2.12/0.9.3/circe-core_2.12-0.9.3.jar
4.0K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/io/circe/circe-parser_2.12/0.9.3/circe-parser_2.12-0.9.3.jar
 76K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/org/spire-math/jawn-parser_2.12/0.11.1/jawn-parser_2.12-0.11.1.jar
172K	/Users/ollie/.coursier/cache/v1/https/oss.sonatype.org/content/repositories/releases/org/scalameta/jsonrpc_2.12/0.2.0/jsonrpc_2.12-0.2.0.jar
3.0M	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/org/typelevel/cats-kernel_2.12/1.1.0/cats-kernel_2.12-1.1.0.jar
 16K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/io/circe/circe-jawn_2.12/0.9.3/circe-jawn_2.12-0.9.3.jar
 12K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/com/outr/perfolation-macros_2.12/1.0.2/perfolation-macros_2.12-1.0.2.jar
292K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/com/outr/scribe_2.12/2.5.0/scribe_2.12-2.5.0.jar
 36K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/org/typelevel/machinist_2.12/0.6.2/machinist_2.12-0.6.2.jar
4.0K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/io/monix/monix_2.12/2.3.3/monix_2.12-2.3.3.jar
4.0M	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/org/typelevel/cats-core_2.12/1.1.0/cats-core_2.12-1.1.0.jar
 16K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/io/circe/circe-numbers_2.12/0.9.3/circe-numbers_2.12-0.9.3.jar
728K	/Users/ollie/.coursier/cache/v1/https/repo1.maven.org/maven2/io/monix/monix-execution_2.12/2.3.3/monix-execution_2.12-2.3.3.jar
 12M	total
```

If you exclude cats the footprint is 4.9mb and if you also exclude monix the footprint is 1.7mb.
